src/code_803D110.c

```diff
@@ -0,0 +1,1091 @@
+            for (k = 0; k < r6; k++) {
+                sp575[j] = currSp;
+                if (++j >= 576)
+                    break;
+            }
+        }
+        else {
+            for (k = 0; k < r6; k++) {
+                currSp = sp[spId++];
+                sp575[j] = currSp;
+                if (++j >= 576)
+                    break;
+            }
+        }
+        if (j >= 576)
+            break;
+    }
+
+    for (i = 0; i < 576; i += 8) {
+        u32 val = sp575[i + 7] << 0x1C;
+        val |= (sp575[i + 6] << 0x18) & 0xF000000;
+        val |= (sp575[i + 5] << 0x14) & 0xF00000;
+        val |= (sp575[i + 4] << 0x10) & 0xF0000;
+        val |= (sp575[i + 3] << 0xC) & 0xF000;
+        val |= (sp575[i + 2] << 0x8) & 0xF00;
+        val |= (sp575[i + 1] << 0x4) & 0xF0;
+        val |= (sp575[i]) & 0xF;
+        *(a1++) = val;
+    }
+}
+
+// Could be a misnomer?
+s32 CalcStringWidth(const u8 *dungName)
+{
+    s32 w = 0;
+
+    while (*dungName != '\0') {
+        u32 currChar = *(dungName++);
+        if (currChar == '#') {
+            w += 8;
+        }
+        else if (currChar == ' ') {
+            w += 12;
+        }
+        else {
+            struct UnkStruct_sub_803DC6C *strPtr;
+
+            if (currChar & 0x80 && *dungName != '\0') {
+                currChar = (currChar << 8) | *dungName;
+                dungName++;
+            }
+            strPtr = sub_803DEC8(currChar);
+            if (strPtr != NULL) {
+                w += strPtr->unk6;
+            }
+        }
+    }
+
+    return w;
+}
+
+struct UnkStruct_sub_803DC6C *sub_803DEC8(s32 chr)
+{
+    s32 r2, r4;
+    struct UnkStruct_sub_803DC6C *ret;
+    struct UnkStruct_sub_803DC6C *strPtr = *((struct UnkStruct_sub_803DC6C **) gDungeonNameBannerFontFile->data);
+    // Fakematch? Or just magic numbers which will make more sense once this file is documented?
+    if (chr > 63487 && chr < 65535)
+    {
+        s32 r2 = chr & 0xFF;
+        s32 r1 = (chr & 0xFF00) / 256;
+        r2 -= 32;
+        r1 -= 248;
+        ret = &strPtr[r1 * 224 + r2];
+    }
+    else
+    {
+        r4 = 0;
+        r2 = gDungeonNameBannerFont - 1;
+        while (r4 < r2) {
+            s32 r1 = (r4 + r2) / 2;
+            if (strPtr[r1].unk4 == chr) {
+                r4 = r1;
+                break;
+            }
+            else if (strPtr[r1].unk4 < chr) {
+                r4 = r1 + 1;
+            }
+            else {
+                r2 = r1;
+            }
+        }
+
+        ret = &strPtr[r4];
+        if (ret->unk4 != chr)
+            ret = &strPtr[1];
+
+
+    }
+    return ret;
+};
+
+/*
+I'm leaving this as it is, because I'm not sure how to generate the ldmia asm instruction. Maybe once the file's data format is better understood.
+struct FileSub_803DF60Struct
+{
+    s32 count;
+    void *unkPtr;
+};
+
+struct OpenedFile_sub_803DF60
+{
+    File *file;
+    struct FileSub_803DF60Struct **strPtr;
+};
+
+void sub_803DF60(void)
```

Comment:
Is there an in-progress decomp.me link we can add here?

Reply:
I didn't make any